### ISSUE TITLE
Independent replication: re-measure exp199 contacts-v1 rollout R-precision (tie-break for #209)

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#212]
Tracking issue for an **independent, uncontaminated replication** of the contact
R-precision measurement for the exp199 contacts-v1 checkpoint, requested off the
`REPLICATION_PROMPT.md` written for #208 and disputed in #209.

Two prior measurements of the *same* checkpoint disagree — **0.5873** vs
**0.6103** (mean R-precision, `all` range) — a gap ~10x the 0.0023 reproducibility
span of this recipe (#204). This issue tracks a third measurement, run on
independent hardware without reading #209's analysis first, to break the tie.

* Source prompt: [`experiments/exp208_models_rl_post_training_on_contacts_v1/REPLICATION_PROMPT.md`](https://github.com/Open-Athena/MarinFold/blob/claude/marinfold-experiment-plan-3ef077/experiments/exp208_models_rl_post_training_on_contacts_v1/REPLICATION_PROMPT.md) @ branch `claude/marinfold-experiment-plan-3ef077`
* Related: #209 (the disagreement), #208 (which the prompt was written for), #199 (the checkpoint), #180 (the frontier tracker)

Result is filed as a comment below.

---

## The request, as given

> Run this analysis and report your results: https://github.com/Open-Athena/MarinFold/blob/claude/marinfold-experiment-plan-3ef077/experiments/exp208_models_rl_post_training_on_contacts_v1/REPLICATION_PROMPT.md

<details>
<summary><b>The full replication prompt it points to (verbatim)</b></summary>

# Task: independently re-measure contact R-precision for the exp199 contacts-v1 model

We have two disagreeing numbers for the **same** checkpoint and need a third,
independent measurement. Please run this yourself and report the number you get.
Do not read our analysis first — we want an uncontaminated replication.

Context: MarinFold issue #209 (https://github.com/Open-Athena/MarinFold/issues/209).

## What to measure

Mean **R-precision, "all" range (sequence separation >= 6)**, over the fixed
554-protein contact eval set, using the **rollout-consensus** inference recipe.

## The checkpoint — pick ONE and say which

All three carry **byte-identical weights** (LFS sha256 `e8db3b66…` for shard 1,
`a7a38503…` for shard 2). They differ only in `config.json` / `tokenizer_config.json`:

| | location | rope stated as | `tokenizer_class` |
|---|---|---|---|
| A | `open-athena/marinfold-exp199` @ revision `ed7103bfd7dac3f75ba759e5ec827da3d75ff0ed`, subfolder `prot-exp199-cw-cv1-s02-m1-p06-aug/hf/step-145199` | `rope_parameters` only | `TokenizersBackend` |
| B | `hf://buckets/open-athena/MarinFold/checkpoints/prot-exp199-cw-cv1-s02-m1-p06-aug/hf/step-145199` (the `MODELS.yaml` default) | `rope_theta` + `rope_scaling` **and** `rope_parameters` | `TokenizersBackend` |
| C | `timodonnell/marinfold-contacts-v1-exp199-1_5b-step145199` | both forms | `PreTrainedTokenizerFast` |

**This matters and is the most likely source of a spurious discrepancy:**

- If your stack runs **transformers 5.x**, A / B / C all load correctly.
- If it runs **transformers 4.x**, A and B fail: 4.x ignores `rope_parameters`
  and *silently* loads `rope_theta = 10000` against a model trained at **500000**
  (wrong, no error), and `TokenizersBackend` is unresolvable so
  `AutoTokenizer.from_pretrained` raises outright.

**Before generating anything, print and report:** your transformers version, the
resolved `rope_theta`, `vocab_size` (expect **2845**), and the ids of
`<contact>`, `<begin_statements>`, `<end>`, `<p0>` (expect **5, 9, 10, 143**).
If `rope_theta` is not 500000, stop — the run would be meaningless.

Weights must be **bfloat16** at load (`dtype="bfloat16"`). The A/B copies are
fp32 on disk; casting at load is fine and is what vLLM does anyway.

## The eval code — use exactly this, do not reimplement

Repo `Open-Athena/MarinFold`, branch `main` (pin commit
`dd7670d2ecdaa415d9f23fac536075a4aa1bde4e`):

1. `experiments/exp82_evals_contacts_v1_contact_prediction/score_rollout_worker.py`
   — generation + vote matrices. Runs unmodified on CUDA and on TPU; all I/O is
   fsspec, so `s3://`, `gs://` and local paths all work. Shard it however suits
   your hardware (`--shard i/n`); it is resumable by `(dataset, stem)`.
2. `.../fetch_cw_scores.py` — sparse vote parquets -> `[L,L]` npz score matrices.
   It enforces 554 units and refuses any protein appearing in more than one part.
3. `.../build_rollout_rows.py` — per-protein metrics. This carries exp89's
   `compute_metrics` verbatim; **do not substitute another metric implementation.**

**There is an `eval-checkpoint` skill in the repo. Use it for the surrounding
discipline (checkpoint identity, compute/storage locality, provenance records) but
NOT for the scorer**: that skill specifies exp89's *pairwise* scorer, which
measures a different quantity and lands near 0.33. We need the *rollout* recipe
above, which lands near 0.6.

## Exact recipe

```
--n-rollouts 100 --temperature 1.0 --top-p 0.95 --top-k -1 --contact-mult 6 --seed 0
```

Add `--no-per-request-seed` **only if** you are on TPU (the JAX backend rejects
per-request seeds). Please report which you used, and your `--max-num-seqs`
(the worker's default is 512; the run we are trying to reproduce used **128**).

**Do not change `--top-p`.** This is a replication: both numbers under dispute
were measured at 0.95, and every published MarinFold rollout number uses it, so
it is what makes your result comparable to them.

For context, since 0.95 is a convention rather than a tuned optimum and you may
wonder: the only sweep behind it (exp82, `data/results_rollout_sweep_dev.txt`)
covered **16 dev proteins** and read, in long-band R-precision — pairwise
reference 0.160; T1.0/p0.95 **0.189**; T0.7/p0.95 **0.192**; T0.7/p1.0 **0.191**;
T1.0/p1.0 **0.184**. The p=0.95 and p=1.0 rows are within noise of each other at
that sample size, and the sweep ran `top_k 50`, which the settled recipe later
dropped to off (#142 traced under-generation to a finite top_k) — so no row in it
is exactly the recipe in use today. The robust signal there is rollout >>
pairwise, not the specific p. Whether 0.95 is the right operating point is a
worthwhile separate question and wants its own sweep on all 554 proteins; it is
explicitly **not** what this task is for.

## The prompts are resampled per rollout — this is not a knob, and not a variance source

`score_rollout_worker.py` builds a **fresh serialization of the protein for every
rollout**:

```python
doc = build_document(f"{stem}:r{k}", residues, [], config=GenerationConfig())
```

`build_document` is pure and deterministic given its first argument, which is the
RNG seed, and the seed differs per rollout `k`. Two things vary across the 100:

1. **Position-token offset** — `start = rng.randrange(num_indices)`, so the `<pN>`
   numbering is rotated differently each time. Absolute position tokens carry no
   fixed meaning across rollouts.
2. **Sequence-statement order** — `rng.shuffle(seq_statements)` shuffles the
   per-residue assignments and both termini.

So the 100 rollouts are 100 *different serializations of the same protein*, not
100 samples from one prompt. That resampling is the diversity the consensus vote
is built on. The worker keeps a per-rollout `{position_token -> seq_index}` map
and decodes every prediction back to sequence indices before voting, so votes
accumulate in a common frame.

Consequence worth knowing: prompt construction is **fully determined by the
stem**, so it is reproducible across runs and machines and cannot be a source of
disagreement between your number and ours. The only sampling randomness is the
engine seed.

## Fixed inputs — fetch, do not rebuild

- Ground truth universe:
  `https://huggingface.co/buckets/open-athena/MarinFold/resolve/data/contacts-v1-model-eval-exp169/gt_universe.jsonl`
  Verify **554** `(dataset, stem)` units and **552** unique stems.
- Eval targets parquet (`dataset`/`stem`/`L`/`input_seq`), whichever is local to you:
  - TPU / GCS: `gs://marin-us-central1/protein-structure/MarinFold/exp169/eval_targets.parquet`
  - CoreWeave / S3: `s3://marin-us-east-02a/MarinFold/exp167_eval/eval_targets.parquet`

## Steps

1. Preflight the checkpoint (the printouts listed above).
2. Smoke-test one shard with `--limit 2` end to end before the full fan-out.
3. Generate all 554 proteins x 100 rollouts.
4. `fetch_cw_scores.py --parts <out>/<label> --out <matrices>` — must report
   `completeness OK: 554/554`.
5. `build_rollout_rows.py --gt gt_universe.jsonl --model <label>=<matrices>
   --out rows.csv.gz --summary summary.csv`
6. Report the mean of `precision` over rows with `cut == "R"` and
   `range == "all"` (n=554), and the same for `range == "long"` (n=553).
   Please also attach `rows.csv.gz` so we can compare per protein rather than
   only on the mean.

## What we expect, and what would be informative

Two prior measurements of this checkpoint disagree: **0.5873** and **0.6103**
(all-range), both at the recipe above. The gap is ~10x the reproducibility span of
this recipe — four evaluations of one unchanged checkpoint span 0.0023 — so your
number should land clearly on one side rather than in between.

Please report the number **before** looking at issue #209's analysis, and include:
transformers version, resolved `rope_theta`, `vocab_size`, the four token ids,
accelerator, `--max-num-seqs`, per-request-seed setting, and which checkpoint copy
(A/B/C) you used.

</details>

**Comment by timodonnell** [comment]
## Result: **0.6110** — this supports 0.6103, not 0.5873

Mean R-precision, `cut == "R"`, over the fixed 554-protein eval set at the rollout-consensus recipe.

| range | n | valid | **mean** | median | SEM |
|---|---|---|---|---|---|
| **all** | 554 | 554 | **0.611032** | 0.6721 | 0.0103 |
| **long** | 554 | 553 | **0.564536** | 0.6463 | 0.0115 |

0.6110 is **0.0007 from 0.6103** — under a third of the 0.0023 reproducibility span (#204) — and **0.0237 from 0.5873**, ~10x that span. It does not land in between. (`long` has 553 valid: one protein has no long-range candidate pair, as the prompt anticipated.)

Full cut x range table:

| cut | all | long | medium | short |
|---|---|---|---|---|
| **R** | **0.6110** | **0.5645** | 0.6580 | 0.6791 |
| L | 0.5510 | 0.3635 | 0.1831 | 0.1204 |
| L/2 | 0.7147 | 0.5611 | 0.3393 | 0.2300 |
| L/5 | 0.8197 | 0.7084 | 0.6177 | 0.4797 |
| AUC | 0.9487 | 0.9347 | 0.9685 | 0.9693 |

R-precision (all) by source dataset: `denovo_pdb` 0.6734 (n=396), `foldbench100` 0.5410 (n=100), `cameo_hard` 0.3829 (n=32), `casp_fm` 0.2114 (n=26).

## Configuration, as requested

| | |
|---|---|
| checkpoint copy | **A** — `open-athena/marinfold-exp199` @ `ed7103bfd7dac3f75ba759e5ec827da3d75ff0ed`, subfolder `prot-exp199-cw-cv1-s02-m1-p06-aug/hf/step-145199` |
| shard sha256 | `e8db3b66…` / `a7a38503…` — recomputed on download, matches |
| transformers | 5.15.0 |
| vLLM / torch | 0.19.1 / 2.10.0+cu128 |
| resolved `rope_theta` | **500000** |
| `vocab_size` | **2845** |
| token ids | `<contact>` 5, `<begin_statements>` 9, `<end>` 10, `<p0>` 143 |
| accelerator | 1x NVIDIA RTX A5000 (24 GB), driver 535.54.03 / CUDA 12.2 |
| `--max-num-seqs` | **128** |
| per-request seed | **enabled** (CUDA backend, so `--no-per-request-seed` not needed) |

Recipe exactly as specified — `--n-rollouts 100 --temperature 1.0 --top-p 0.95 --top-k -1 --contact-mult 6 --seed 0` — with `--shard 0/1`, one process, `--gpu-frac 0.90` and `--chunk 8` (worker defaults). Eval code unmodified from `main` @ `dd7670d2ecdaa415d9f23fac536075a4aa1bde4e`: `score_rollout_worker.py` -> `fetch_cw_scores.py` -> `build_rollout_rows.py`. The `eval-checkpoint` skill was used for discipline only, not for the scorer.

## Run integrity

- 554 x 100 rollouts in **100.0 min**, single process, no preemption or resume.
- **unfinished 0/55400 (0.00%)** — every rollout terminated on `<end>`; none hit the `6L + 128` cap.
- `fetch_cw_scores.py`: `completeness OK: 554/554 (dataset, stem) units, 552 unique stems`, no protein in more than one part.
- `build_rollout_rows.py`: `scored 554/554`, 11080 rows = 554 x 20.
- Inputs fetched, not rebuilt. `gt_universe.jsonl`: 554 units / 552 unique stems. `eval_targets.parquet` (from `gs://marin-us-central1/...`): 554 rows, `(dataset, stem)` set identical to the GT universe, every `L` matches, `sum(L) = 102554`, `L` in [30, 761].
- Gated on a `--limit 2` smoke run end to end (generation -> matrices -> metrics) before the full fan-out.

## Three things from setup worth recording

**1. The rope is llama3-*scaled*, not plain — checking `rope_theta` alone would not catch a mismatch.** All three copies state `rope_parameters = {rope_type: llama3, rope_theta: 500000, factor: 8.0, low_freq_factor: 1.0, high_freq_factor: 4.0, original_max_position_embeddings: 8192}`. I verified the frequencies two independent ways rather than trusting the key:

- transformers' `ROPE_INIT_FUNCTIONS["llama3"]` on this config divides the low-frequency components by 8 (max deviation 8.9e-04 from plain theta=500000 rope), i.e. the scaling really is applied;
- vLLM independently builds `Llama3RotaryEmbedding(base=500000, scaling_factor=8.0)`, and its `cos_sin_cache` agrees with transformers' `inv_freq` to **4.1e-08** while differing from a plain theta=10000 rope by **1.3e-01**.

So the kernel that ran carries the trained rope. Separately: transformers warns that `original_max_position_embeddings` (8192) must be *less than* `max_position_embeddings` (8192) — the config is self-inconsistent by that check, though it loads and scales anyway. Probably worth a look independent of this measurement.

**2. "All three carry byte-identical weights" is not literally true of copy C.** A/B are fp32 (4.98 GB + 906 MB, hashes as published); C is bf16 (2.49 GB + 453 MB, `7d3e7725…` / `7a25ff36…`), so its LFS hashes cannot match. The substantive claim does hold and was checked directly: all **267/267** tensors satisfy `bf16(A) == C` **exactly**. Since the worker loads at `dtype="bfloat16"` either way, A and C are the same weights in GPU memory — this number is equally a measurement of C. (C's `tokenizer_class` is `PreTrainedTokenizerFast` as documented; A's is `TokenizersBackend`.)

**3. Why vLLM 0.19.1.** The workstation driver is CUDA 12.2. Current vLLM (0.27.1) pins torch 2.13 and vLLM 0.20–0.26 pin torch 2.11 — both **CUDA 13** builds that refuse to initialise on a 12.x driver. vLLM 0.19.1 is the newest release on torch 2.10 (`cu12`), and its constraint (`transformers != 5.0.*…!= 5.5.0, >= 4.56.0`) still admits transformers 5.x. That is what made copy A directly loadable instead of forcing the transformers-4.x fallback to C. Anyone reproducing this on an older-driver box will want that pin.

## Artifacts

Published to the public HF bucket under `results/exp212_exp199_rprecision_replication/` (anonymously readable):

- [`rows.csv.gz`](https://huggingface.co/buckets/open-athena/MarinFold/resolve/results/exp212_exp199_rprecision_replication/rows.csv.gz) — per-protein rows in exp89 schema, 11080 rows, for per-protein comparison rather than only on the mean
- [`summary.csv`](https://huggingface.co/buckets/open-athena/MarinFold/resolve/results/exp212_exp199_rprecision_replication/summary.csv) — model x range x cut means
- [`PROVENANCE.md`](https://huggingface.co/buckets/open-athena/MarinFold/resolve/results/exp212_exp199_rprecision_replication/PROVENANCE.md) — full record of the above

Reported without reading #209's analysis.
